components/server/src/oauth-server/oauth-authorization-server.ts

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { AuthorizationRequest, AuthorizationServer, DateInterval, GrantIdentifier, JwtService, OAuthAuthCodeRepository, OAuthClient, OAuthTokenRepository, OAuthUserRepository, RequestInterface, ResponseInterface } from "@jmondi/oauth2-server";
+import {
+  inMemoryClientRepository,
+  inMemoryScopeRepository
+} from "./repository";
+
+const clientRepository = inMemoryClientRepository;
+const scopeRepository = inMemoryScopeRepository;
+
+// TODO(rl) - get this from external secret
+const jwtService = new JwtService("secret secret secret");
```

Comment:
please don't forget about this TODO. 
is it correct that this service is used to produce the issued codes in the end? see: https://github.com/jasonraimondi/typescript-oauth2-server/blob/153c83bd64940a03ba31c833d4e3b2aeb4ce03cd/src/grants/auth_code.grant.ts#L57

Reply:
This is less important than it might appear. 

The service is used to encrypt and [decrypt](https://github.com/jasonraimondi/typescript-oauth2-server/blob/5b60bc0d4ad41a19dd6033e4d242b7db13203f58/src/grants/auth_code.grant.ts#L55) the auth code which is returned untouched by the client and the value is only used internally. In saying that, I agree it should be taken from an external source to make it less likely that external parties can create a valid auth code. Importantly, PKCE is the key here - it might be a valid code but they won't have the valid challenge/verifier that is required alongside it.

On the client side, the jwtService secret is used for _validation_ of the returned token to detect MITM among others, and involves a shared secret between the client and the server. The actual token can be decoded and used without knowing the secret (e.g. see the [online decoder](http://calebb.net/). Given that we are communicating with only our server and that having a shared secret in public code for the local app client is not secret, then @csweichel and I decided that we can safely ignore this validation step.

Reply:
I think you are ignoring the PKCE-side - this is why that standard was created.

Later on, it checks the [code challenge](https://github.com/jasonraimondi/typescript-oauth2-server/blob/153c83bd64940a03ba31c833d4e3b2aeb4ce03cd/src/grants/auth_code.grant.ts#L79) against the original request and will reject it.

Reply:
@rl-gitpod, indeed it's following the validation.

the thing is, the `authCode.codeChallenge` is looked up by `validatedPayload.auth_code_id` which – as mentioned in 2.) can be forged from a valid authorization request made in parallel, just because it's then easy to decode and recompose to a new one.

Reply:
Sorry, I missed what you meant in 2 - long day.

if it is a publicly known secret I agree, but as it will be coming from an external source this isn't an issue.

Reply:
that's why I commented with the reminder
> please don't forget about this TODO.

Reply:
I reckon we should make this come from an external source prior to merging the PR. Pushing this to `env.ts` and the `values.yaml` isn't much of an effort.

Reply:
> I agree it should be taken from an external source 
That was the plan...

Done... from the logs
> 2021-05-20T04:44:40.948Z.  JWT:rdqfUiocN2Ql7fYdFKDN

It is random, but as the auth code is very short-lived it it shouldn't be a problem if it is redeployed (unless you are very unlucky).

I'll circle back to reducing the logging later, but now I've given into the virus that I've been fighting with the last couple of days.